proposals/0000-type-indexed-typeable.rst

```diff
@@ -0,0 +1,488 @@
+
+.. [PeytonJones2016]
+    Peyton Jones, Weirich, Eisenberg, Vytiniotis. "`A Reflection on Types
+    <http://research.microsoft.com/en-us/um/people/simonpj/papers/haskell-dynamic/index.htm>`__."
+    *Proc. Philip Wadler's 60th birthday Festschrift*. Edinburgh, April 2016.
+
+Like today, the new ``Typeable`` mechanism will only support kind-monomorphic
+types. Unlike today's mechanism, we provide a means of extracting the *kind* of
+a type representation,
+
+.. code-block:: haskell
+
+    -- | The kind of a type.
+    typeRepKind :: TypeRep (a :: k) -> TypeRep k
+
+With a ``Typeable`` constraint we can get a ``TypeRep`` for a (non-kind
+polymorphic) type ``a`` with ``typeRep``,
+
+.. code-block:: haskell
+
+    class Typeable (a :: k)
+
+    typeRep :: forall (a :: k). Typeable a => TypeRep a
+
+Note how, in contrast to ``Data.Typeable.typeRep``, we needn't provide a
+``Proxy`` to ``typeRep``; the desired type propagates through ``TypeRep``\'s
+index.
+
+
+We can pattern match on the structure of a ``TypeRep``. For instance, on type
+constructors,
+
+.. code-block:: haskell
+
+    -- | A type constructor type. This must be a unidirectional pattern since
+    -- we have no way of preventing the user from instantiating a TyCon in an
+    -- ill-kinded manner.
+    pattern TRCon :: forall k (a :: k). TyCon -> TypeRep a
+
+    -- | Information about a type constructor. No means of constructing 'TyCon's
+    -- is provided; the only values of this type available are those from
+    -- 'TypeRep's.
+    data TyCon
+    tyConPackage :: TyCon -> String
+    tyConModule :: TyCon -> String
+    tyConName :: TyCon -> String
+
+Type application can also be decomposed,
+
+.. code-block:: haskell
+
+    -- | A representation of a type application, @a b@. This is a bidirectional pattern.
+    pattern TRApp :: forall k2 (fun :: k2). ()
+                  => forall k1 (a :: k1 -> k2) (b :: k1). (fun ~ a b)
+                  => TypeRep a -> TypeRep b -> TypeRep fun
+
+We can also decompose function types (e.g. ``Int -> String``) in their argument
+(e.g. ``Int``) and result types (``String``). Strictly speaking this can be
+expressed in terms of ``TRApp`` but it seems like a common enough pattern that
+it's worth providing a pattern for it,
+
+.. code-block:: haskell
+
+    pattern TRFun :: forall fun. ()
+                  => forall arg res. (fun ~ (arg -> res))
+                  => TypeRep arg
+                  -> TypeRep res
+                  -> TypeRep fun
+
+We can also test for type equality,
+
+.. code-block:: haskell
+
+    -- | Kind-homogeneous type equality
+    eqTypeRep  :: forall k (a :: k) (b :: k).
+                  TypeRep a -> TypeRep b -> Maybe (a :~: b)
+
+    -- | Kind-heterogeneous type equality
+    eqTypeRep' :: forall k1 k2 (a :: k1) (b :: k2).
+                  TypeRep a -> TypeRep b -> Maybe (a :~~: b)
+
+    -- | Kind-heterogeneous type equality
+    data (a :: k1) :~~: (b :: k2) where
+        HRefl :: a :~~: a
+
```

Comment:
Nice. Consider adding a heterogenous version of `class TestEquality` to go with your `:~~:`.